File: src/stores/history.ts
/** @fileoverview Pinia store for SQLite-backed download history persistence.
 *
 * Stores completed/errored download records independently from the aria2
 * session file (which only tracks active/paused tasks). Records survive
 * app restarts and upgrades.
 *
 * Database: sqlite:history.db (managed by tauri-plugin-sql with migrations).
 */
import { defineStore } from 'pinia'
import { ref } from 'vue'
import Database from '@tauri-apps/plugin-sql'
import { remove } from '@tauri-apps/plugin-fs'
import { invoke } from '@tauri-apps/api/core'
import { appDataDir } from '@tauri-apps/api/path'
import { collectTaskIdentityBuckets } from '@shared/utils/task'
import type { Aria2Task, HistoryRecord } from '@shared/types'
import { logger } from '@shared/logger'

const DB_NAME = 'sqlite:history.db'
export type HistoryRecordSortField = 'name' | 'status' | 'total_length' | 'task_type' | 'completed_at'
export type HistoryRecordSortOrder = 'ascend' | 'descend' | false

export interface HistoryRecordsPageInput {
  status?: string
  page: number
  pageSize: number
  sortField?: string
  sortOrder?: HistoryRecordSortOrder
}

export interface HistoryRecordsPage {
  records: HistoryRecord[]
  total: number
}

const HISTORY_SORT_COLUMNS: Record<HistoryRecordSortField, string> = {
  name: 'name',
  status: 'status',
  total_length: 'total_length',
  task_type: 'task_type',
  completed_at: 'completed_at',
}

function normalizePage(value: number): number {
  return Math.max(1, Math.floor(Number.isFinite(value) ? value : 1))
}

function normalizePageSize(value: number): number {
  return Math.min(Math.max(1, Math.floor(Number.isFinite(value) ? value : 50)), 100)
}

function resolveHistoryOrderBy(sortField?: string, sortOrder?: HistoryRecordSortOrder): string {
  if (sortOrder && sortField && sortField in HISTORY_SORT_COLUMNS) {
    const column = HISTORY_SORT_COLUMNS[sortField as HistoryRecordSortField]
    return `ORDER BY ${column} ${sortOrder === 'descend' ? 'DESC' : 'ASC'}, COALESCE(added_at, completed_at) DESC`
  }
  return 'ORDER BY COALESCE(added_at, completed_at) DESC'
}

function appendInClause(clauses: string[], params: string[], expression: string, values: string[]): void {
  if (values.length === 0) return
  const placeholders = values.map((_, i) => `$${params.length + i + 1}`).join(', ')
  clauses.push(`${expression} IN (${placeholders})`)
  params.push(...values)
}

/** Callbacks for database health events — allows UI layer to show toasts
 *  without coupling the store to any specific UI framework. */
export interface DbHealthCallbacks {
  onCorrupt?: () => void
  onError?: (error: unknown) => void
  onRebuilt?: () => void
  onRebuildFailed?: (error: unknown) => void
}

export const useHistoryStore = defineStore('history', () => {
  let db: Awaited<ReturnType<typeof Database.load>> | null = null
  let initPromise: Promise<void> | null = null
  const recordTotal = ref(0)

  /** Apply SQLite PRAGMA optimizations to an open connection. */
  async function applyPragmas(conn: NonNullable<typeof db>): Promise<void> {
    await conn.execute('PRAGMA journal_mode = WAL', [])
    await conn.execute('PRAGMA synchronous = NORMAL', [])
    await conn.execute('PRAGMA busy_timeout = 5000', [])
    await conn.execute('PRAGMA foreign_keys = ON', [])
  }

  /** Delete the database files from disk (db + WAL + SHM). */
  async function deleteDbFiles(): Promise<void> {
    try {
      const dataDir = await appDataDir()
      const suffixes = ['history.db', 'history.db-wal', 'history.db-shm']
      for (const suffix of suffixes) {
        const path = `${dataDir}/${suffix}`
        if (await invoke<boolean>('check_path_exists', { path })) {
          await remove(path)
        }
      }
    } catch (e) {
      logger.warn('HistoryDB', `deleteDbFiles failed: ${e}`)
    }
  }

  /** Attempt to rebuild the database from scratch after corruption. */
  async function rebuildDatabase(callbacks?: DbHealthCallbacks): Promise<void> {
    try {
      if (db) {
        try {
          await db.close()
        } catch (e) {
          logger.debug('HistoryDB', `close before rebuild failed (already broken): ${e}`)
        }
        db = null
      }
      await deleteDbFiles()
      db = await Database.load(DB_NAME)
      await applyPragmas(db)
      logger.info('HistoryDB', 'Database rebuilt successfully')
      callbacks?.onRebuilt?.()
    } catch (e) {
      logger.error('HistoryDB', `Rebuild failed: ${e}`)
      db = null
      initPromise = null
      callbacks?.onRebuildFailed?.(e)
    }
  }

  /** Initialize the database connection, verify integrity, and auto-recover
   *  from corruption. Safe to call multiple times — subsequent calls are no-ops.
   *
   *  @param callbacks Optional UI notification hooks for health events. */
  async function init(callbacks?: DbHealthCallbacks): Promise<void> {
    if (db) return
    if (!initPromise) {
      initPromise = (async () => {
        try {
          db = await Database.load(DB_NAME)
          await applyPragmas(db)

          // Verify structural integrity on every cold start
          const result = await db.select<{ integrity_check: string }[]>('PRAGMA integrity_check', [])
          const status = result[0]?.integrity_check ?? 'unknown'
          if (status !== 'ok') {
            logger.warn('HistoryDB', `Integrity check failed: ${status}`)
            callbacks?.onCorrupt?.()
            await rebuildDatabase(callbacks)
          }
          if (db) await refreshRecordTotal()
        } catch (e) {
          logger.warn('HistoryDB', `Init failed: ${e}`)
          callbacks?.onError?.(e)
          await rebuildDatabase(callbacks)
          if (db) await refreshRecordTotal()
        }
      })()
    }
    await initPromise
  }

  /** Returns the active database connection, auto-initializing if needed. */
  async function getDb() {
    if (!db) await init()
    return db!
  }

  async function refreshRecordTotal(): Promise<number> {
    const rows = await (
      await getDb()
    ).select<Array<{ count: number }>>('SELECT COUNT(*) as count FROM download_history', [])
    const total = Number(rows[0]?.count ?? 0)
    recordTotal.value = Number.isFinite(total) ? Math.max(0, total) : 0
    return recordTotal.value
  }

  async function countRecordsMatchingTaskIdentities(tasks: Aria2Task[]): Promise<number> {
    const identities = collectTaskIdentityBuckets(tasks)
    const clauses: string[] = []
    const params: string[] = []

    appendInClause(clauses, params, 'gid', identities.gids)
    appendInClause(clauses, params, "json_extract(meta, '$.infoHash')", identities.btInfoHashes)
    appendInClause(clauses, params, "json_extract(meta, '$.ed2kHash')", identities.ed2kHashes)
    appendInClause(clauses, params, "json_extract(meta, '$.ed2kLink')", identities.ed2kLinks)

    if (clauses.length === 0) return 0

    const rows = await (
      await getDb()
    ).select<Array<{ count: number }>>(
      `SELECT COUNT(DISTINCT gid) as count FROM download_history WHERE ${clauses.join(' OR ')}`,
      params,
    )
    const total = Number(rows[0]?.count ?? 0)
    return Number.isFinite(total) ? Math.max(0, total) : 0
  }

  /** Insert or update a download record (upsert by GID).
   *
   *  Uses ON CONFLICT instead of INSERT OR REPLACE to preserve
   *  the immutable added_at timestamp. REPLACE = DELETE + INSERT
   *  which would reset added_at; ON CONFLICT DO UPDATE keeps it. */
  async function addRecord(record: HistoryRecord): Promise<void> {
    await (
      await getDb()
    ).execute(
      `INSERT INTO download_history
        (gid, name, uri, dir, total_length, status, task_type, added_at, completed_at, meta)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10)
       ON CONFLICT(gid) DO UPDATE SET
         name = excluded.name,
         uri = excluded.uri,
         dir = excluded.dir,
         total_length = excluded.total_length,
         status = excluded.status,
         task_type = excluded.task_type,
         added_at = COALESCE(download_history.added_at, excluded.added_at),
         completed_at = excluded.completed_at,
         meta = excluded.meta`,
      [
        record.gid,
        record.name,
        record.uri ?? null,
        record.dir ?? null,
        record.total_length ?? null,
        record.status,
        record.task_type ?? null,
        record.added_at ?? null,
        record.completed_at ?? null,
        record.meta ?? null,
      ],
    )
    await refreshRecordTotal()
  }

  /** Retrieve records, optionally filtered by status and/or limited in count.
   *  Sorted by added_at DESC for position-stable ordering.
   *  Falls back to completed_at DESC for records without added_at (pre-migration). */
  async function getRecords(status?: string, limit?: number): Promise<HistoryRecord[]> {
    // Normalize limit: floor → clamp to [0, 10000] → append only if finite
    const limitClause = limit != null ? ` LIMIT ${Math.min(Math.max(0, Math.floor(limit)), 10_000)}` : ''
    const orderBy = 'ORDER BY COALESCE(added_at, completed_at) DESC'
    if (status) {
      return (await getDb()).select<HistoryRecord[]>(
        `SELECT * FROM download_history WHERE status = $1 ${orderBy}${limitClause}`,
        [status],
      )
    }
    return (await getDb()).select<HistoryRecord[]>(`SELECT * FROM download_history ${orderBy}${limitClause}`, [])
  }

  async function getRecordsPage(input: HistoryRecordsPageInput): Promise<HistoryRecordsPage> {
    const page = normalizePage(input.page)
    const pageSize = normalizePageSize(input.pageSize)
    const offset = (page - 1) * pageSize
    const orderBy = resolveHistoryOrderBy(input.sortField, input.sortOrder)
    const limitClause = `LIMIT ${pageSize} OFFSET ${offset}`
    const conn = await getDb()

    if (input.status) {
      const [records, countRows] = await Promise.all([
        conn.select<HistoryRecord[]>(`SELECT * FROM download_history WHERE status = $1 ${orderBy} ${limitClause}`, [
          input.status,
        ]),
        conn.select<Array<{ count: number }>>('SELECT COUNT(*) as count FROM download_history WHERE status = $1', [
          input.status,
        ]),
      ])
      return { records, total: Number(countRows[0]?.count ?? 0) }
    }

    const [records, countRows] = await Promise.all([
      conn.select<HistoryRecord[]>(`SELECT * FROM download_history ${orderBy} ${limitClause}`, []),
      conn.select<Array<{ count: number }>>('SELECT COUNT(*) as count FROM download_history', []),
    ])
    return { records, total: Number(countRows[0]?.count ?? 0) }
  }

  /** Retrieve a single record by GID, or null if not found. */
  async function getRecordByGid(gid: string): Promise<HistoryRecord | null> {
    const rows = await (
      await getDb()
    ).select<HistoryRecord[]>('SELECT * FROM download_history WHERE gid = $1 LIMIT 1', [gid])
    return rows[0] ?? null
  }

  /** Remove a single record by GID. */
  async function removeRecord(gid: string): Promise<void> {
    await (await getDb()).execute('DELETE FROM download_history WHERE gid = $1', [gid])
    await refreshRecordTotal()
  }

  /** Remove task birth timestamps for the provided GIDs. */
  async function removeBirthRecords(gids: string[]): Promise<void> {
    if (gids.length === 0) return
    const placeholders = gids.map((_, i) => `$${i + 1}`).join(', ')
    await (await getDb()).execute(`DELETE FROM task_birth WHERE gid IN (${placeholders})`, gids)
  }

  /** Remove all records, optionally filtered by status. Full reset also VACUUMs. */
  async function clearRecords(status?: string): Promise<void> {
    if (status) {
      await (await getDb()).execute('DELETE FROM download_history WHERE status = $1', [status])
    } else {
      await (await getDb()).execute('DELETE FROM download_history', [])
      // VACUUM reclaims disk space and resets AUTOINCREMENT counter
      await (await getDb()).execute('VACUUM', [])
    }
    await refreshRecordTotal()
  }

  /** Remove records whose GIDs are in the provided list (stale file cleanup). */
  async function removeStaleRecords(gids: string[]): Promise<void> {
    if (gids.length === 0) return
    const placeholders = gids.map((_, i) => `$${i + 1}`).join(', ')
    await (await getDb()).execute(`DELETE FROM download_history WHERE gid IN (${placeholders})`, gids)
    await refreshRecordTotal()
  }

  /** Remove records matching a BT infoHash stored in the meta JSON column.
   *
   * Aria2 reassigns GIDs on session restore, so the same torrent may have
   * stale DB records under old GIDs. This method cleans them up before
   * writing a fresh record with the current GID, preventing duplicates.
   *
   * Optionally excludes a specific GID to avoid deleting the record
   * that was just written (INSERT OR REPLACE is by GID, not infoHash). */
  async function removeByInfoHash(infoHash: string, excludeGid?: string): Promise<void> {
    if (!infoHash) return
    if (excludeGid) {
      await (
        await getDb()
      ).execute(`DELETE FROM download_history WHERE json_extract(meta, '$.infoHash') = $1 AND gid != $2`, [
        infoHash,
        excludeGid,
      ])
    } else {
      await (
        await getDb()
      ).execute(`DELETE FROM download_history WHERE json_extract(meta, '$.infoHash') = $1`, [infoHash])
    }
    await refreshRecordTotal()
  }

  /** Run PRAGMA integrity_check and return the result string. */
  async function checkIntegrity(): Promise<string> {
    const result = await (await getDb()).select<{ integrity_check: string }[]>('PRAGMA integrity_check', [])
    return result[0]?.integrity_check ?? 'unknown'
  }

  /** Close the database connection and reset initialization state.
   *  After calling, the next init() or getDb() will re-open the database. */
  async function closeConnection(): Promise<void> {
    if (db) {
      await db.close()
      db = null
    }
    initPromise = null
  }

  /** Persist task birth timestamp to the task_birth table.
   *  INSERT OR IGNORE ensures the first write wins — added_at is immutable. */
  async function recordTaskBirth(gid: string, addedAt?: string): Promise<void> {
    await (
      await getDb()
    ).execute(`INSERT OR IGNORE INTO task_birth (gid, added_at) VALUES ($1, $2)`, [
      gid,
      addedAt ?? new Date().toISOString(),
    ])
  }

  /** Load all birth records from the task_birth table.
   *  Called on app startup to pre-populate the in-memory addedAtMap. */
  async function loadBirthRecords(): Promise<Array<{ gid: string; added_at: string }>> {
    return (await getDb()).select<Array<{ gid: string; added_at: string }>>('SELECT gid, added_at FROM task_birth', [])
  }

  /** Query the DB schema version from tauri_plugin_sql's internal tracking table. */
  async function getSchemaVersion(): Promise<number> {
    try {
      const result = await (
        await getDb()
      ).select<Array<{ version: number }>>('SELECT MAX(version) as version FROM _sqlx_migrations', [])
      return result[0]?.version ?? 0
    } catch (e) {
      logger.debug('HistoryDB', `schema version query failed: ${e}`)
      return 0
    }
  }

  return {
    init,
    addRecord,
    getRecords,
    getRecordsPage,
    recordTotal,
    refreshRecordTotal,
    countRecordsMatchingTaskIdentities,
    getRecordByGid,
    removeRecord,
    removeBirthRecords,
    clearRecords,
    removeStaleRecords,
    removeByInfoHash,
    checkIntegrity,
    closeConnection,
    recordTaskBirth,
    loadBirthRecords,
    getSchemaVersion,
  }
})

File: src-tauri/src/history.rs
//! History database repository — Rust-side access to `history.db`.
//!
//! Provides the same CRUD operations as the frontend `useHistoryStore` but
//! accessible from Rust commands. Uses `rusqlite` directly (NOT through
//! tauri-plugin-sql) to avoid IPC round-trips for backend consumers like
//! stale-record cleanup and task lifecycle monitor.
//!
//! The database schema is still managed by tauri-plugin-sql migrations —
//! this module only reads/writes to existing tables.

use crate::error::AppError;
use rusqlite::{params, Connection, OptionalExtension};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::sync::Arc;
use tokio::sync::Mutex;

/// Rust-side mirror of the TypeScript `HistoryRecord` interface.
///
/// Field names use `snake_case` to match the SQLite column names.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryRecord {
    pub id: Option<i64>,
    pub gid: String,
    pub name: String,
    pub uri: Option<String>,
    pub dir: Option<String>,
    pub total_length: Option<i64>,
    pub status: String,
    pub task_type: Option<String>,
    pub added_at: Option<String>,
    pub created_at: Option<String>,
    pub completed_at: Option<String>,
    pub meta: Option<String>,
}

/// Thread-safe history database handle.
///
/// Uses `Mutex<Connection>` because `rusqlite::Connection` is `!Send` on
/// some configurations.  The mutex is uncontended in practice — backend
/// writes are infrequent and read-only queries are fast.
pub struct HistoryDb {
    conn: Arc<Mutex<Connection>>,
}

/// Tauri managed state wrapper.
pub struct HistoryDbState(pub Arc<HistoryDb>);

impl HistoryDb {
    /// Opens (or creates) the history database at the given path.
    ///
    /// Applies WAL journal mode and busy timeout PRAGMAs matching the
    /// frontend's `applyPragmas()` function.
    pub fn open(path: &PathBuf) -> Result<Self, AppError> {
        let conn = Connection::open(path)?;
        conn.execute_batch(
            "PRAGMA journal_mode = WAL;
             PRAGMA synchronous = NORMAL;
             PRAGMA busy_timeout = 5000;
             PRAGMA foreign_keys = ON;",
        )?;
        Ok(Self {
            conn: Arc::new(Mutex::new(conn)),
        })
    }

    /// Opens an in-memory database for testing, with the schema pre-applied.
    #[cfg(test)]
    pub fn open_in_memory() -> Result<Self, AppError> {
        let conn = Connection::open_in_memory()?;
        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS download_history (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                gid TEXT NOT NULL UNIQUE,
                name TEXT NOT NULL,
                uri TEXT,
                dir TEXT,
                total_length INTEGER,
                status TEXT NOT NULL DEFAULT 'complete',
                task_type TEXT,
                added_at TEXT,
                created_at TEXT DEFAULT (datetime('now')),
                completed_at TEXT,
                meta TEXT
             );
             CREATE TABLE IF NOT EXISTS task_birth (
                gid TEXT PRIMARY KEY,
                added_at TEXT NOT NULL
             );",
        )?;
        Ok(Self {
            conn: Arc::new(Mutex::new(conn)),
        })
    }

    /// Upsert a history record by GID.
    ///
    /// Uses ON CONFLICT(gid) DO UPDATE to preserve the immutable `added_at`.
    /// Matches the frontend's `addRecord()` SQL exactly.
    pub async fn add_record(&self, record: &HistoryRecord) -> Result<(), AppError> {
        let conn = self.conn.lock().await;
        conn.execute(
            "INSERT INTO download_history
                (gid, name, uri, dir, total_length, status, task_type, added_at, completed_at, meta)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10)
             ON CONFLICT(gid) DO UPDATE SET
                name = excluded.name,
                uri = excluded.uri,
                dir = excluded.dir,
                total_length = excluded.total_length,
                status = excluded.status,
                task_type = excluded.task_type,
                added_at = COALESCE(download_history.added_at, excluded.added_at),
                completed_at = excluded.completed_at,
                meta = excluded.meta",
            params![
                record.gid,
                record.name,
                record.uri,
                record.dir,
                record.total_length,
                record.status,
                record.task_type,
                record.added_at,
                record.completed_at,
                record.meta,
            ],
        )?;
        Ok(())
    }

    /// Query records, optionally filtered by status.
    ///
    /// Sorted by `COALESCE(added_at, completed_at) DESC` matching the frontend.
    pub async fn get_records(
        &self,
        status: Option<&str>,
        limit: Option<u32>,
    ) -> Result<Vec<HistoryRecord>, AppError> {
        let conn = self.conn.lock().await;
        let order = "ORDER BY COALESCE(added_at, completed_at) DESC";
        let limit_clause = limit
            .map(|l| format!(" LIMIT {}", l.min(10_000)))
            .unwrap_or_default();

        if let Some(status) = status {
            let sql =
                format!("SELECT * FROM download_history WHERE status = ?1 {order}{limit_clause}");
            let mut stmt = conn.prepare(&sql)?;
            let rows = stmt.query_map(params![status], Self::row_to_record)?;
            let records: Vec<HistoryRecord> = rows.collect::<Result<Vec<_>, _>>()?;
            Ok(records)
        } else {
            let sql = format!("SELECT * FROM download_history {order}{limit_clause}");
            let mut stmt = conn.prepare(&sql)?;
            let rows = stmt.query_map([], Self::row_to_record)?;
            let records: Vec<HistoryRecord> = rows.collect::<Result<Vec<_>, _>>()?;
            Ok(records)
        }
    }

    /// Remove a single record by GID.
    pub async fn remove_record(&self, gid: &str) -> Result<(), AppError> {
        let conn = self.conn.lock().await;
        conn.execute("DELETE FROM download_history WHERE gid = ?1", params![gid])?;
        Ok(())
    }

    /// Clear all records, optionally filtered by status.
    pub async fn clear_records(&self, status: Option<&str>) -> Result<(), AppError> {
        let conn = self.conn.lock().await;
        if let Some(status) = status {
            conn.execute(
                "DELETE FROM download_history WHERE status = ?1",
                params![status],
            )?;
        } else {
            conn.execute("DELETE FROM download_history", [])?;
            conn.execute_batch("VACUUM")?;
        }
        Ok(())
    }

    /// Remove records whose GIDs are in the provided list (stale cleanup).
    pub async fn remove_stale_records(&self, gids: &[String]) -> Result<(), AppError> {
        if gids.is_empty() {
            return Ok(());
        }
        let conn = self.conn.lock().await;
        let placeholders: Vec<String> = gids
            .iter()
            .enumerate()
            .map(|(i, _)| format!("?{}", i + 1))
            .collect();
        let sql = format!(
            "DELETE FROM download_history WHERE gid IN ({})",
            placeholders.join(", ")
        );
        let params: Vec<&dyn rusqlite::ToSql> =
            gids.iter().map(|g| g as &dyn rusqlite::ToSql).collect();
        conn.execute(&sql, params.as_slice())?;
        Ok(())
    }

    /// Record a task birth timestamp (INSERT OR IGNORE — first write wins).
    pub async fn record_task_birth(&self, gid: &str, added_at: &str) -> Result<(), AppError> {
        let conn = self.conn.lock().await;
        conn.execute(
            "INSERT OR IGNORE INTO task_birth (gid, added_at) VALUES (?1, ?2)",
            params![gid, added_at],
        )?;
        Ok(())
    }

    /// Return the first recorded birth timestamp for a task GID.
    pub async fn get_task_birth(&self, gid: &str) -> Result<Option<String>, AppError> {
        let conn = self.conn.lock().await;
        let added_at = conn
            .query_row(
                "SELECT added_at FROM task_birth WHERE gid = ?1",
                params![gid],
                |row| row.get(0),
            )
            .optional()?;
        Ok(added_at)
    }

    /// Load all birth records for pre-populating in-memory maps.
    pub async fn load_birth_records(&self) -> Result<Vec<(String, String)>, AppError> {
        let conn = self.conn.lock().await;
        let mut stmt = conn.prepare("SELECT gid, added_at FROM task_birth")?;
        let records = stmt
            .query_map([], |row| Ok((row.get(0)?, row.get(1)?)))?
            .collect::<Result<Vec<_>, _>>()?;
        Ok(records)
    }

    /// Check database integrity.
    pub async fn check_integrity(&self) -> Result<String, AppError> {
        let conn = self.conn.lock().await;
        let result: Option<String> = conn
            .query_row("PRAGMA integrity_check", [], |row| row.get(0))
            .optional()?;
        Ok(result.unwrap_or_else(|| "unknown".to_string()))
    }

    /// Map a rusqlite Row to a HistoryRecord.
    fn row_to_record(row: &rusqlite::Row) -> rusqlite::Result<HistoryRecord> {
        Ok(HistoryRecord {
            id: row.get("id")?,
            gid: row.get("gid")?,
            name: row.get("name")?,
            uri: row.get("uri")?,
            dir: row.get("dir")?,
            total_length: row.get("total_length")?,
            status: row.get("status")?,
            task_type: row.get("task_type")?,
            added_at: row.get("added_at")?,
            created_at: row.get("created_at")?,
            completed_at: row.get("completed_at")?,
            meta: row.get("meta")?,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_record(gid: &str, name: &str, status: &str) -> HistoryRecord {
        HistoryRecord {
            id: None,
            gid: gid.to_string(),
            name: name.to_string(),
            uri: Some("http://example.com/file.zip".to_string()),
            dir: Some("/tmp".to_string()),
            total_length: Some(1024),
            status: status.to_string(),
            task_type: Some("uri".to_string()),
            added_at: Some("2025-01-01T00:00:00Z".to_string()),
            created_at: None,
            completed_at: Some("2025-01-01T01:00:00Z".to_string()),
            meta: None,
        }
    }

    // ── CRUD operations ─────────────────────────────────────────────

    #[tokio::test]
    async fn add_and_get_record() {
        let db = HistoryDb::open_in_memory().unwrap();
        let rec = make_record("gid001", "test.zip", "complete");
        db.add_record(&rec).await.unwrap();

        let records = db.get_records(None, None).await.unwrap();
        assert_eq!(records.len(), 1);
        assert_eq!(records[0].gid, "gid001");
        assert_eq!(records[0].name, "test.zip");
        assert_eq!(records[0].status, "complete");
        assert_eq!(records[0].total_length, Some(1024));
    }

    #[tokio::test]
    async fn upsert_preserves_added_at() {
        let db = HistoryDb::open_in_memory().unwrap();
        let rec1 = HistoryRecord {
            added_at: Some("2025-01-01T00:00:00Z".to_string()),
            ..make_record("gid001", "test.zip", "active")
        };
        db.add_record(&rec1).await.unwrap();

        // Upsert with a different added_at — COALESCE should preserve the original
        let rec2 = HistoryRecord {
            added_at: Some("2025-06-01T00:00:00Z".to_string()),
            status: "complete".to_string(),
            ..make_record("gid001", "test-updated.zip", "complete")
        };
        db.add_record(&rec2).await.unwrap();

        let records = db.get_records(None, None).await.unwrap();
        assert_eq!(records.len(), 1);
        assert_eq!(records[0].name, "test-updated.zip");
        assert_eq!(records[0].status, "complete");
        // added_at preserved from first insert (COALESCE keeps original)
        assert_eq!(records[0].added_at.as_deref(), Some("2025-01-01T00:00:00Z"));
    }

    #[tokio::test]
    async fn get_records_filters_by_status() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.add_record(&make_record("g1", "a.zip", "complete"))
            .await
            .unwrap();
        db.add_record(&make_record("g2", "b.zip", "error"))
            .await
            .unwrap();
        db.add_record(&make_record("g3", "c.zip", "complete"))
            .await
            .unwrap();

        let complete = db.get_records(Some("complete"), None).await.unwrap();
        assert_eq!(complete.len(), 2);

        let errors = db.get_records(Some("error"), None).await.unwrap();
        assert_eq!(errors.len(), 1);
        assert_eq!(errors[0].gid, "g2");
    }

    #[tokio::test]
    async fn get_records_respects_limit() {
        let db = HistoryDb::open_in_memory().unwrap();
        for i in 0..10 {
            let mut rec = make_record(&format!("g{i}"), &format!("file{i}.zip"), "complete");
            rec.added_at = Some(format!("2025-01-{:02}T00:00:00Z", i + 1));
            db.add_record(&rec).await.unwrap();
        }

        let limited = db.get_records(None, Some(3)).await.unwrap();
        assert_eq!(limited.len(), 3);
    }

    #[tokio::test]
    async fn remove_record_by_gid() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.add_record(&make_record("g1", "a.zip", "complete"))
            .await
            .unwrap();
        db.add_record(&make_record("g2", "b.zip", "complete"))
            .await
            .unwrap();

        db.remove_record("g1").await.unwrap();

        let records = db.get_records(None, None).await.unwrap();
        assert_eq!(records.len(), 1);
        assert_eq!(records[0].gid, "g2");
    }

    #[tokio::test]
    async fn clear_records_by_status() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.add_record(&make_record("g1", "a.zip", "complete"))
            .await
            .unwrap();
        db.add_record(&make_record("g2", "b.zip", "error"))
            .await
            .unwrap();

        db.clear_records(Some("error")).await.unwrap();

        let records = db.get_records(None, None).await.unwrap();
        assert_eq!(records.len(), 1);
        assert_eq!(records[0].gid, "g1");
    }

    #[tokio::test]
    async fn clear_all_records() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.add_record(&make_record("g1", "a.zip", "complete"))
            .await
            .unwrap();
        db.add_record(&make_record("g2", "b.zip", "error"))
            .await
            .unwrap();

        db.clear_records(None).await.unwrap();

        let records = db.get_records(None, None).await.unwrap();
        assert!(records.is_empty());
    }

    #[tokio::test]
    async fn remove_stale_records() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.add_record(&make_record("g1", "a.zip", "complete"))
            .await
            .unwrap();
        db.add_record(&make_record("g2", "b.zip", "complete"))
            .await
            .unwrap();
        db.add_record(&make_record("g3", "c.zip", "complete"))
            .await
            .unwrap();

        db.remove_stale_records(&["g1".to_string(), "g3".to_string()])
            .await
            .unwrap();

        let records = db.get_records(None, None).await.unwrap();
        assert_eq!(records.len(), 1);
        assert_eq!(records[0].gid, "g2");
    }

    #[tokio::test]
    async fn remove_stale_records_empty_is_noop() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.add_record(&make_record("g1", "a.zip", "complete"))
            .await
            .unwrap();
        db.remove_stale_records(&[]).await.unwrap();
        assert_eq!(db.get_records(None, None).await.unwrap().len(), 1);
    }

    // ── Task birth tracking ─────────────────────────────────────────

    #[tokio::test]
    async fn record_and_load_task_births() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.record_task_birth("g1", "2025-01-01T00:00:00Z")
            .await
            .unwrap();
        db.record_task_birth("g2", "2025-01-02T00:00:00Z")
            .await
            .unwrap();

        let births = db.load_birth_records().await.unwrap();
        assert_eq!(births.len(), 2);
    }

    #[tokio::test]
    async fn record_task_birth_ignores_duplicate() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.record_task_birth("g1", "2025-01-01T00:00:00Z")
            .await
            .unwrap();
        // Second insert should be silently ignored
        db.record_task_birth("g1", "2025-06-01T00:00:00Z")
            .await
            .unwrap();

        let births = db.load_birth_records().await.unwrap();
        assert_eq!(births.len(), 1);
        assert_eq!(births[0].1, "2025-01-01T00:00:00Z");
    }

    #[tokio::test]
    async fn get_task_birth_returns_persisted_added_at() {
        let db = HistoryDb::open_in_memory().unwrap();
        db.record_task_birth("g1", "2025-01-01T00:00:00Z")
            .await
            .unwrap();

        assert_eq!(
            db.get_task_birth("g1").await.unwrap().as_deref(),
            Some("2025-01-01T00:00:00Z")
        );
        assert_eq!(db.get_task_birth("missing").await.unwrap(), None);
    }

    // ── Integrity check ─────────────────────────────────────────────

    #[tokio::test]
    async fn check_integrity_returns_ok() {
        let db = HistoryDb::open_in_memory().unwrap();
        let result = db.check_integrity().await.unwrap();
        assert_eq!(result, "ok");
    }

    // ── Serialization contract ──────────────────────────────────────

    #[test]
    fn history_record_serializes_to_camel_case_free_json() {
        // HistoryRecord fields are snake_case matching SQL columns.
        // This test ensures serde doesn't silently rename anything.
        let rec = make_record("g1", "test.zip", "complete");
        let json = serde_json::to_value(&rec).unwrap();
        assert!(json.get("gid").is_some());
        assert!(json.get("total_length").is_some());
        assert!(json.get("added_at").is_some());
        assert!(json.get("task_type").is_some());
        // No camelCase — frontend expects snake_case from SQL columns
        assert!(json.get("totalLength").is_none());
        assert!(json.get("addedAt").is_none());
    }
}
